### PR TITLE
Add support for constructing expressions from string and evaluating

### PR DESCRIPTION
**Pull request by save-buffer**
This PR adds support for constructing Velox expression objects in Python from a string, and allows for evaluating the expression on Velox vectors in python.

### CODE REVIEW

**Comment by kagamiori** [comment]
Hi Sasha, Could you add a context about this PR in the description? Thanks!

**Comment by save-buffer** [comment]
@kagamiori Added a short description

File: pyvelox/pyvelox.h [hunk 1]
```diff
@@ -366,11 +400,77 @@ inline void addVectorBindings(
...
+            for (IExprWrapper& w : i)
+              inputs.push_back(w.expr);
+            IExprWrapper result = {e.expr->withInputs(inputs)};
+            return result;
+          })
+      .def(
+          "evaluate",
+          [](IExprWrapper& e,
+             std::vector<std::string> names,
+             std::vector<VectorPtr>& inputs) {
```
Comment (kgpai) [inline pyvelox/pyvelox.h:439]: Can we have evaluate that takes a map from string-> VectorPtr , so we can have expressions like : 
```
  result =  Expression.from_string("a + b")
                .evaluate({ "a": a, "b":b}) 
  ```

File: pyvelox/pyvelox.h [hunk 2]
```diff
@@ -366,11 +400,77 @@ inline void addVectorBindings(
...
+            if (names.size() != inputs.size()) {
+              throw py::value_error(
+                  "Must specify the same number of names as inputs");
+            }
+            vector_size_t numRows = inputs.empty() ? 0 : inputs[0]->size();
+            std::vector<std::shared_ptr<const Type>> types;
+            types.reserve(inputs.size());
+            for (auto v : inputs) {
```
Comment (kgpai) [inline pyvelox/pyvelox.h:447]: nit: can you give a name like vector instead of v.

File: pyvelox/test/test_expression.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import unittest
+
+
+class TestVeloxExpression(unittest.TestCase):
+    def test_from_string(self):
+        pv.Expression.from_string("a + b")
+
+    def test_eval(self):
+        expr = pv.Expression.from_string("a + b")
+        a = pv.from_list([1, 2, 3, None])
```
Comment (kgpai) [inline pyvelox/test/test_expression.py:25]: Can you add some test expressions with : 
 1. a function
 2. expression with cast
 3. provide more arguments than required
 4. vectors with all nulls
 5. expression that works on strings, ints, doubles (to have some sort of exhaustive checks)
 6. a somewhat nested expression
 7. expression with switch/case

File: velox/duckdb/memory/CMakeLists.txt [hunk 4]
```diff
@@ -14,7 +14,8 @@
 
 add_library(velox_duckdb_allocator Allocator.cpp)
 
-target_link_libraries(velox_duckdb_allocator velox_dwio_common duckdb ${FMT})
+target_link_libraries(velox_duckdb_allocator velox_dwio_common duckdb ${FMT}
+                      ${FOLLY_WITH_DEPENDENCIES})
```
Comment (kgpai) [inline velox/duckdb/memory/CMakeLists.txt:18]: Why do we have folly_with_dependencies here ?

**Review by kgpai — COMMENTED** [review]
Looks great, excited about the progress !

File: pyvelox/pyvelox.h [hunk 5]
```diff
@@ -29,6 +34,36 @@ namespace facebook::velox::py {
 
 namespace py = pybind11;
 
+struct PyVeloxContext {
+  static inline PyVeloxContext& getInstance() {
+    if (!instance_)
+      instance_ = std::make_unique<PyVeloxContext>();
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:40]: Added

File: pyvelox/pyvelox.h [hunk 6]
```diff
@@ -366,11 +400,77 @@ inline void addVectorBindings(
...
+  py::class_<IExprWrapper>(
+      m, "Expression", py::module_local(asModuleLocalDefinitions))
+      .def("__str__", [](IExprWrapper& e) { return e.expr->toString(); })
+      .def("getInputs", [](IExprWrapper& e) { return e.expr->getInputs(); })
+      .def(
+          "withInputs",
+          [](IExprWrapper& e, std::vector<IExprWrapper>& i) {
+            std::vector<std::shared_ptr<const core::IExpr>> inputs;
+            inputs.reserve(i.size());
+            for (IExprWrapper& w : i)
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:430]: Added

File: pyvelox/pyvelox.h (hunk 2 above)
Comment (save-buffer) [reply pyvelox/pyvelox.h:447]: Changed

File: pyvelox/pyvelox.h [hunk 7]
```diff
@@ -366,11 +400,77 @@ inline void addVectorBindings(
...
+  // PyBind11's classes cannot be const, but the parse functions return const
+  // shared_ptrs, so we wrap in a non-const class.
+  struct IExprWrapper {
+    std::shared_ptr<const core::IExpr> expr;
+  };
+
+  py::class_<IExprWrapper>(
+      m, "Expression", py::module_local(asModuleLocalDefinitions))
+      .def("__str__", [](IExprWrapper& e) { return e.expr->toString(); })
+      .def("getInputs", [](IExprWrapper& e) { return e.expr->getInputs(); })
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:424]: Added doc strings

File: pyvelox/pyvelox.h (hunk 1 above)
Comment (save-buffer) [reply pyvelox/pyvelox.h:439]: Added

**Comment by save-buffer** [comment]
Not super sure what's going on with CI:
```
/opt/rh/gcc-toolset-9/root/bin/g++  -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CONTEXT_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DFMT_LOCALE -DGFLAGS_IS_A_DLL=0 -DVELOX_DISABLE_GOOGLETEST -I../../. -I../../velox/external/xxhash -I_deps/folly-src -I_deps/folly-build -I/usr/include/libdwarf -I../../third_party/xsimd/include -isystem ../../velox -isystem ../../velox/external -isystem ../../velox/external/duckdb -isystem ../../velox/external/duckdb/tpch/dbgen/include -mavx2 -mfma -mavx -mf16c -mlzcnt -std=c++17 -mbmi2 -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -g -fPIC   -Wno-deprecated-declarations -Wno-stringop-overflow -std=gnu++17 -MD -MT velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ComparisonFunctionsRegistration.cpp.o -MF velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ComparisonFunctionsRegistration.cpp.o.d -o velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ComparisonFunctionsRegistration.cpp.o -c ../../velox/functions/prestosql/registration/ComparisonFunctionsRegistration.cpp
g++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
```

**Review by kgpai — APPROVED** [review]

File: CMakeLists.txt [hunk 8]
```diff
@@ -353,6 +353,7 @@ if(NOT ${VELOX_BUILD_MINIMAL})
   endif()
 endif()
 
+message("FOLLY: ${FOLLY_INCLUDE_DIRS}")
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply CMakeLists.txt]: Oh yes, sorry that was left over from debugging

**Pull request closed without merging**